File: 48.py
#!/usr/bin/env python

"""This script solves the Project Euler problem "Self powers, and
hexagonal". The problem is: Find the last ten digits of the series,
1^1 + 2^2 + 3^3 + ... + 1000^1000.
"""

from __future__ import division


def main():
    """Self powers"""

    # Constants
    NUM_DIGITS = 10
    LAST_NUMBER = 1000

    sum_end = 0
    for number in range(1, LAST_NUMBER + 1):
        power_end = number
        for _ in range(number - 1):
            power_end *= number
            power_end = int(str(power_end)[-NUM_DIGITS:])
        sum_end += power_end
        sum_end = int(str(sum_end)[-NUM_DIGITS:])

    print("{:010d}".format(sum_end))

if __name__ == '__main__':
    main()

File: 48.js
#!/usr/bin/env node

// Self powers

var NUM_DIGITS = 10;
var LAST_NUMBER = 1000;

var sum_end = 0;
for (var number = 1; number <= LAST_NUMBER; number++) {
    var power_end = number;
    for (var i = 1; i < number; i++) {
        power_end *= number;
        power_end = parseInt(power_end.toString().slice(-NUM_DIGITS));
    }
    sum_end += power_end;
    sum_end = parseInt(sum_end.toString().slice(-NUM_DIGITS));
}
sum_end = ('0000000000' + sum_end).slice(-NUM_DIGITS);

console.log(sum_end);
